src/bin/elfshaker/status.rs

```diff
@@ -0,0 +1,172 @@
+
+    let idx = repo.load_index(snapshot.pack())?;
+    let handles = idx
+        .resolve_snapshot(snapshot.tag())
+        .expect("failed to resolve snapshot"); // TODO: Temporary.
+
+    for entry in idx.entries_from_handles(handles.iter())? {
+        let path = Path::new(&entry.path);
+
+        let changed = if path.exists() == false {
+            // missing in workspace
+            info!("not in workspace {}", path.display());
+            true
+        } else {
+            let workspace_is_symlink = path.is_symlink();
+
+            if entry.file_metadata.is_symlink_file != workspace_is_symlink {
+                info!(
+                    "symlink status differs from recording for {}",
+                    path.display()
+                );
+                true
+            } else {
+                if workspace_is_symlink {
+                    let workspace_target = fs::read_link(path)?;
+                    let changed = entry.file_metadata.symlink_target != workspace_target;
+                    if changed {
+                        info!(
+                            "changed \"{}\": index.symlink: {}; fs.symlink: {};",
+                            path.display(),
+                            entry.file_metadata.symlink_target.display(),
+                            workspace_target.display()
+                        );
+                        true
+                    } else {
+                        info!(
+                            "same \"{}\": symlink_target: {};",
+                            path.display(),
+                            workspace_target.display()
+                        );
+                        false
+                    }
+                } else {
+                    let workspace_checksum = calculate_sha1(&path)?;
+
+                    if entry.checksum != workspace_checksum {
+                        info!(
+                            "changed \"{}\": index.checksum: {}; fs.checksum: {};",
+                            path.display(),
+                            hex::encode(entry.checksum),
+                            hex::encode(workspace_checksum)
+                        );
+                        true
+                    } else {
+                        info!(
+                            "same \"{}\": checksum: {};",
+                            path.display(),
+                            hex::encode(workspace_checksum)
+                        );
+                        false
+                    }
+                }
+            }
+        };
+
+        if changed {
+            changed_files.push(path.display().to_string());
+        }
+    }
+
+    changed_files.sort();
+
+    Ok(changed_files)
+}
+
+fn calculate_sha1(path: &Path) -> std::io::Result<[u8; 20]> {
+    let mut file = open_file(path)?;
+    let size = file.metadata().map(|m| m.len() as usize);
+    let mut buffer = Vec::with_capacity(size.unwrap_or(0));
+
+    file.read_to_end(&mut buffer)?;
```

Comment:
This will crash as well on typical customers build, because you are allocating contiguously file buffers of more than 4GB and often OS can't provide you with some much memory.